### ISSUE TITLE
ParallelUpdater cannot fully exploit multiple GPUs

### ISSUE THREAD

**Issue by iwiwi** [chainer/chainer#1649]
`ParallelUpdater` improves performance when using simple models and/or large mini batches. On the other hand, for complex models, `ParallelUpdater` currently fails to exploit multiple GPUs, and it becomes almost sequential.

I examined the profiler results, and concluded that Python is just too slow to issue kernels for multiple GPUs. The possible way to cope with this issue is to use multiple processes in `ParallelUpdater`. To enable that, cupy needs the feature to handle inter-process communication. It can be done using the inter-process memory handle of CUDA (http://docs.nvidia.com/cuda/cuda-c-programming-guide/#interprocess-communication).

The following is a profiler result with 4 GPUs and ResNet-152. It clearly shows that multiple GPUs are used in sequential.

<img width="1464" alt="2016-08-10 18 39 22" src="https://cloud.githubusercontent.com/assets/469803/18466558/21880536-79d7-11e6-9f8f-ebd117f22d68.png">

**Comment by jekbradbury** [comment]
I have a patch which uses CUDA's interprocess communication and `multiprocessing` to give near-linear speedup for data parallelism across multiple GPUs, but it was written for Chainer 1.5 and I'd have to do some work to bring it up to compatibility with the latest version. Is there interest in this? I'm not sure when I'll get time to work on it, but it'll probably be in the next few weeks.

**Comment by beam2d** [comment]
We are very interested in your code. It is great if you can spend some time to make your code compatible to the current Chainer and send a patch, but if it is difficult just sharing your code as is is still much helpful to us.

**Comment by anaruse** [comment]
Last week, I spend some of my time to see if we can improve chainer's multi-GPU performance with python multiprocessing module. The experimental coding using the multiprocessing module with CUDA IPC APIs has been nearly finished. So, I'm attaching the results with ResNet-152 on 4 GPUs. As you can see from it, multiple GPUs are running in parallel and decent speed-up is achieved even in case of very deep network like ResNet-152.

![screenshot](https://cloud.githubusercontent.com/assets/20274068/20053166/ac2f56d0-a51a-11e6-9f17-3d85ef8fb5d7.png)

If you have interest in it, I will spend some more time to refactor and clean-up the code and send PR.  For now, some features are implemented in kind of not-easy-to-maintain manner...

**Comment by beam2d** [comment]
It looks great! We are very interested in the improvement of parallel learning with better multiprocessing with CUDA IPC APIs. Would you send us a PR?

**Comment by anaruse** [comment]
Thank you for your comment. I've just started code refactoring and clean-up work, and I think I can send you a first PR today or tomorrow if it is OK that the patch is not available with chainer's trainer module. Now, I'm testing multi-GPUs training w/o trainer.

**Comment by beam2d** [comment]
Great, thank you. I think there is no problem to first merge the code w/o supporting trainer.

**Comment by jekbradbury** [comment]
I spent some time looking through the patch I wrote earlier; it was fairly specific to the task I was using it for (machine translation) and would need a lot of work to make it general enough. It sounds like what anaruse has is more helpful, but I'm happy to help in any way I can. [Here's my old code](https://gist.github.com/jekbradbury/794d7ffeb13a3cf5b7b76bba0a074a8a) in case it's useful to anyone (the only thing this depends on is a wrapper for CUDA's ipcMemHandle API).
